### PR TITLE
double checked locking for AdditionalTextFile.GetText

### PR DESCRIPTION
**Pull request by YairHalberstadt**
Fixes https://github.com/dotnet/roslyn/issues/41301

cc @sharwell

### CODE REVIEW

File: src/Compilers/Core/Portable/AdditionalTextFile.cs [hunk 1]
```diff
@@ -43,18 +43,20 @@ public AdditionalTextFile(CommandLineSourceFile sourceFile, CommonCompiler compi
         /// Returns a <see cref="SourceText"/> with the contents of this file, or <c>null</c> if
         /// there were errors reading the file.
         /// </summary>
-        public override SourceText GetText(CancellationToken cancellationToken = default)
+        public override SourceText? GetText(CancellationToken cancellationToken = default)
         {
-            lock (_lockObject)
+            if (_text == null && _diagnostics == null)
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/AdditionalTextFile.cs]: what's the current status of double-checked locking.  Is it guaranteed to be safe on .net fx/core? Does it work with these complex check patterns?

**Comment by YairHalberstadt** [comment]
@cyrusNajmabadi

I've force pushed to remove unwanted changes.

File: src/Compilers/Core/Portable/AdditionalTextFile.cs (hunk 1 above)
Comment (CyrusNajmabadi) [reply src/Compilers/Core/Portable/AdditionalTextFile.cs]: can we just use Lazy (or LazyInitializer) instead?  I trust the framework far more here than any sort of manual double-checked lock.

**Comment by CyrusNajmabadi** [comment]
> force pushed to remove unwanted changes.

Thanks!

**Comment by YairHalberstadt** [comment]
@CyrusNajmabadi Why?

**Comment by CyrusNajmabadi** [comment]
Because it feels like `Lazy<T>` but with more work :)  In the case where you don't already have a lock for it to use, and you need to add an extra bool, it's just a lot of bookeeping over just Lazy :)

File: src/Compilers/Core/Portable/AdditionalTextFile.cs [hunk 2]
```diff
@@ -17,11 +17,9 @@ internal sealed class AdditionalTextFile : AdditionalText
     {
         private readonly CommandLineSourceFile _sourceFile;
         private readonly CommonCompiler _compiler;
-        private SourceText? _text;
+        private Lazy<SourceText?> _text;
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/AdditionalTextFile.cs]: `readonly`?

**Comment by jcouv** [comment]
Thanks @YairHalberstadt for the contribution!

**Pull request merged**